hoverkite-protocol/src/response.rs

```diff
@@ -0,0 +1,189 @@
+                    side
+                } else {
+                    side.opposite()
+                };
+                let bytes: Vec<u8> = buffer.drain(0..8).collect();
+                let position = i64::from_le_bytes(bytes.try_into().unwrap());
+                Some(Response {
+                    side,
+                    response: SideResponse::Position(position),
+                })
+            } else {
+                None
+            }),
+            Some(b'B') | Some(b'b') => Ok(if buffer.len() >= 7 {
+                let side = if buffer.pop_front().unwrap() == b'B' {
+                    side
+                } else {
+                    side.opposite()
+                };
+                let bytes: Vec<u8> = buffer.drain(0..6).collect();
+                let battery_voltage = u16::from_le_bytes(bytes[0..2].try_into().unwrap());
+                let backup_battery_voltage = u16::from_le_bytes(bytes[2..4].try_into().unwrap());
+                let motor_current = u16::from_le_bytes(bytes[4..6].try_into().unwrap());
+                Some(Response {
+                    side,
+                    response: SideResponse::BatteryReadings {
+                        battery_voltage,
+                        backup_battery_voltage,
+                        motor_current,
+                    },
+                })
+            } else {
+                None
+            }),
+            Some(b'C') | Some(b'c') => Ok(if buffer.len() >= 2 {
+                let side = if buffer.pop_front().unwrap() == b'C' {
+                    side
+                } else {
+                    side.opposite()
+                };
+                let byte = buffer.pop_front().unwrap();
+                let charger_connected = match byte {
+                    b'0' => false,
+                    b'1' => true,
+                    _ => return Err(UnexpectedResponse(byte)),
+                };
+                Some(Response {
+                    side,
+                    response: SideResponse::ChargeState { charger_connected },
+                })
+            } else {
+                None
+            }),
+            Some(r) => {
+                buffer.pop_front();
+                Err(UnexpectedResponse(r))
+            }
+            None => Ok(None),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use test_case::test_case;
+
+    #[test]
+    fn parse_invalid() {
+        let mut buffer = VecDeque::new();
+        buffer.extend(b"x");
+        assert_eq!(
+            Response::parse(&mut buffer, Side::Right),
+            Err(UnexpectedResponse(b'x'))
+        );
+        assert_eq!(buffer.len(), 0);
+    }
+
+    #[test]
+    fn parse_empty() {
+        let mut buffer = VecDeque::new();
+        assert_eq!(Response::parse(&mut buffer, Side::Right), Ok(None));
+        assert_eq!(buffer.len(), 0);
+    }
+
+    // TODO: test_case doesn't allow test cases which differ only in case.
```

Comment:
You probably want to give each of these a name anyway, because the names that it infers from the bye strings don't have anything to do with our domain messages.

Reply:
Done.